CRM/Financial/BAO/FinancialAccount.php

```diff
@@ -366,7 +366,7 @@ public static function validateFinancialAccount($financialAccountId, $financialA
    *
```

Comment:
The logic of this function is off IMHO. 

First, if there is a $contributionID passed in, shouldn't that take precedence over the $params, which appear to be ones that could be passed to create a contribution. In other words, the actual contribution should be treated as more authoritative than the parameters to create it. Please check its values first and ignore params altogether if contribution is passed in. Don't assume inside side function anything about relationship between params and contribution.

Second, iirc when a contribution is created when financial_type is specified for the contribution but not for the line items, then the civicrm_contribution.financial_type_id is used to set the civicrm_line_item.financial_type_id. The values for the line item financial types are more authoritative than civicrm_contribution's, so they should be checked first before the contribution's when checking for deferred revenue accounts that derive from the params. More specifically, if there is no contribution and one ends up checking the params, then one should check if there is a deferred revenue account for any of the line_item and return true if there is. Only if there is at least one line item in params without a financial type specified, should one check if there is a deferred revenue account for the param's contribution.financial_type_id (L385 is thus wrong, since that check should be done after not before checking the line items).